emu_mps/tdvp.py

```diff
@@ -73,8 +73,27 @@ def apply_effective_Hamiltonian(
     # this order seems to be pretty balanced, but needs to be
     # revisited when use-cases are more well-known
     state = torch.tensordot(left_bath, state, 1)
-    state = torch.tensordot(state, ham, ([1, 2], [0, 2]))
-    state = torch.tensordot(state, right_bath, ([3, 1], [1, 2]))
+    state = state.permute(0, 3, 1, 2)
+    ham = ham.permute(0, 2, 1, 3)
+    state = state.reshape(state.shape[0], state.shape[1], -1).contiguous()
+    ham = ham.reshape(-1, ham.shape[2], ham.shape[3]).contiguous()
+    state = torch.tensordot(state, ham, 1)
+    state = state.permute(0, 2, 1, 3)
+    state = state.reshape(state.shape[0], state.shape[1], -1).contiguous()
+    right_bath = right_bath.permute(2, 1, 0)
+    right_bath = right_bath.reshape(-1, right_bath.shape[2]).contiguous()
+    state = torch.tensordot(state, right_bath, 1)
+
+    # Alternatively:
```

Comment:
what's this?

Reply:
let's just leave the one that is not commented out. A year ago, I did a manual computation of the flop counts as they depend on the contraction order, and from that you could extract an optimal order dependent on the dimensions. At the time I figured we could always revisit it later, and by now I've forgotten the details. If you think it's worth the effort we could do this kind of analysis in a separate MR though.

Reply:
done